backend/src/modules/posts/post.service.ts

```diff
@@ -0,0 +1,234 @@
+      where: { user: { id: userId } },
+    });
+    if (!author) {
+      throw new NotFoundException(await this.t('post.author_not_found'));
+    }
+    return author;
+  }
+
+  /** Format tên: chữ đầu mỗi từ viết hoa, xóa khoảng trắng thừa */
+  private formatName(name: string): string {
+    return name
+      .trim()
+      .toLowerCase()
+      .split(/\s+/)
+      .map((word) => word[0].toUpperCase() + word.slice(1))
+      .join(' ');
+  }
+
+  async create(
+    userId: number,
+    dto: CreatePostDto,
+    file?: Express.Multer.File,
+  ): Promise<PostSerializer> {
+    const queryRunner = this.dataSource.createQueryRunner();
+
+    // 1. Kết nối và bắt đầu transaction
+    await queryRunner.connect();
+    await queryRunner.startTransaction();
+
+    try {
+      // 2. Lấy Author
+      const author = await queryRunner.manager.findOne(Author, {
+        where: { user: { id: userId } },
+      });
+      if (!author)
+        throw new NotFoundException(await this.t('post.author_not_found'));
+
+      // 3. Handle Category
+      let category: Category | undefined;
+      if (dto.categoryId) {
+        const foundCategory = await queryRunner.manager.findOne(Category, {
+          where: { id: dto.categoryId },
+        });
+        if (!foundCategory)
+          throw new NotFoundException(await this.t('post.category_not_found'));
+        category = foundCategory;
+      } else if (dto.categoryName) {
+        const formattedName = this.formatName(dto.categoryName);
+
+        const existingCategory = await queryRunner.manager.findOne(Category, {
+          where: { name: formattedName },
+        });
+
+        if (existingCategory) {
+          throw new BadRequestException(
+            await this.t('post.category_already_exists'),
+          );
+        }
+
+        category = queryRunner.manager.create(Category, {
+          name: formattedName,
+        });
+        await queryRunner.manager.save(category);
+      }
+
+      // 4. Handle Tags
+      const tags: Tag[] = [];
+
+      if (dto.tagIds?.length) {
+        const foundTags = await queryRunner.manager.find(Tag, {
+          where: { id: In(dto.tagIds) },
+        });
+        if (foundTags.length !== dto.tagIds.length)
+          throw new NotFoundException(await this.t('post.tag_not_found'));
+        tags.push(...foundTags);
+      }
+
+      if (dto.tagNames?.length) {
+        for (const name of dto.tagNames) {
+          const formattedName = this.formatName(name);
+          const existingTag = await queryRunner.manager.findOne(Tag, {
+            where: { name: formattedName },
+          });
+
+          if (existingTag) {
+            throw new BadRequestException(
+              await this.t('post.tag_already_exists'),
+            );
+          }
+
+          const tag = queryRunner.manager.create(Tag, { name: formattedName });
+          await queryRunner.manager.save(tag);
+          tags.push(tag);
+        }
+      }
```

Comment:
Tách các phần xử lý logic như check and create category, check and create tags, handle images, ...  thành các method private để clean code xử lý, dễ bảo trì e nha